### DISCUSSION TITLE
Flatpak and Bevy

### DISCUSSION THREAD

**Discussion by lukors** [bevyengine/bevy#3793] [General]
Is it possible to Flatpak a Bevy app? Would I want to? Are there any considerations?

I have Flatpaked maybe three GTK apps, so I know a bit about it, but I just followed a tutorial then and don't understand it fully.

The first issue I ran into when looking into Flatpak for Bevy is the application identifier. In GTK you set it when you spawn the window, and I believe a correct identifier is a requirement for it to work properly? **I can't figure out how to set the application identifier when using Bevy.**

I'm not looking for a tutorial as those already exist (though a Bevy-specific one would be a great asset), I'm more interested in Flatpak from the perspective of a Bevy user.

**Comment by heavyrain266** [comment]
I would not recommend using flatpak, snap or any other way to statically package your bevy game and used environment. Program which requires access to native graphics drivers will most likely require including them alongside your statically packaged program which automatically removes ways to use it outside of your own hardware or using it in general since you can't statically link/package drivers.

**Reply by bjorn3** [reply]
If you only care about opensource drivers you can use the freedesktop sdk. It bundles a very recent (most of the time the latest) version of mesa and as flatpak runtime it gets updated independently from your flatpak app.

**Reply by lukors** [reply]
I found [this Flatpak issue](https://github.com/flatpak/flatpak/issues/138) which gives some more information about using the proprietary Nvidia drivers in Flatpaks. Seems like this is not well supported yet, but apparently it should be possible, and some work is going into it.

I don't know anything about the open source (Nouveau) vs proprietary Nvidia drivers so I did some Internet searching. [This article about Nouveau on Phoronix](https://www.phoronix.com/scan.php?page=article&item=nouveau-kepler-2021&num=1) from June 2021 talks about how even almost 10 year old Nvidia GPUs still have really poor performance with Nouveau.

Unfortunately this removes Flatpak as an option for me. I'll look into this again when they have solved the issue above, so people can use the proprietary driver in Flatpaks. Thanks a lot to both of you for the information that allowed me to reach a decision! :D

**Reply by eugineerd** [reply]
I'm not really familiar with how to package flatpaks but [veloren](https://gitlab.com/veloren/veloren) seems to work fine in it. Both nvidia and intel work for me just fine, maybe you can find something useful in their [flathub manifest](https://github.com/flathub/net.veloren.veloren)?

**Reply by lukors** [reply]
@eugineerd that's interesting, thanks a lot! :smile: 

After more searching based on your message it seems like there is a way to use Nvidia proprietary drivers in Flatpak seemingly easily—though I haven't found out how yet, and it seems like it comes with some issues ([source 1](https://github.com/flatpak/flatpak/issues/2718), [source 2](https://github.com/flathub/org.freedesktop.Platform.GL.nvidia/issues/42)).

I don't have an Nvidia GPU on Linux to test how this works. So unless I get my hands on that I'm not going to risk giving a bad experience to the Nvidia + Linux people.

If someone has a link to documentation about how all this works, please post it! With good enough information I'd be comfortable going down the Flatpak route even if I can't test it myself.

**Reply by zicklag** [reply]
I don't have a link to good documentation, but as a Linux user with Nvidia graphics, I haven't noticed issues using Flatpaks like Blender or Veloren, and I've noticed that flatpak does install and update Nvidia drivers of some sort, and they are updated almost every time I bother running `flatpak update`.

I'm on a Pop!_OS machine with the proprietary Nvidia drivers installed natively, though that probably doesn't make a difference for Flatpaks.

Maybe this doesn't shed much light on anything, but it looks like it might be possible to usefully distribute graphics apps with Flatpak even with Nvidia.

My card: NVIDIA GeForce GTX 1650
